cdap-app-fabric/src/main/java/io/cdap/cdap/security/auth/AuditLogSubscriberService.java

```diff
@@ -0,0 +1,168 @@
+      cConf.getInt(Constants.AuditLogging.AUDIT_LOG_FETCH_SIZE),
+      cConf.getInt(TxConstants.Manager.CFG_TX_TIMEOUT),
+      cConf.getInt(Constants.AuditLogging.AUDIT_LOG_POLL_DELAY_MILLIS),
+      RetryStrategies.exponentialDelay(10, 200, TimeUnit.MILLISECONDS),
+      metricsCollectionService.getContext(ImmutableMap.of(
+        Constants.Metrics.Tag.COMPONENT, Constants.Service.MASTER_SERVICES,
+        Constants.Metrics.Tag.INSTANCE_ID, "0",
+        Constants.Metrics.Tag.NAMESPACE, NamespaceId.SYSTEM.getNamespace(),
+        Constants.Metrics.Tag.TOPIC, cConf.get(Constants.AuditLogging.AUDIT_LOG_EVENT_TOPIC),
+        Constants.Metrics.Tag.CONSUMER, Constants.AuditLogging.AUDIT_LOG_CONSUMER_WRITER_SUBSCRIBER
+      )));
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.transactionRunner = transactionRunner;
+    this.accessControllerInstantiator = accessControllerInstantiator;
+  }
+
+  /**
+   * Returns the {@link TransactionRunner} for executing tasks in transaction.
+   */
+  @Override
+  protected TransactionRunner getTransactionRunner() {
+    return transactionRunner;
+  }
+
+  /**
+   * Loads last persisted message id. This method will be called from a transaction. The returned
+   * message id will be used as the starting message id (exclusive) for the first fetch.
+   */
+  @Nullable
+  @Override
+  protected String loadMessageId(StructuredTableContext context) throws Exception {
+    AppMetadataStore appMetadataStore = AppMetadataStore.create(context);
+    return appMetadataStore.retrieveSubscriberState(getTopicId().getTopic(),
+                                                    Constants.AuditLogging.AUDIT_LOG_WRITER_SUBSCRIBER);
+  }
+
+  /**
+   * Persists the given message id. This method will be called from a transaction, which is the same
+   * transaction for the call to {@link #processMessages(StructuredTableContext, Iterator)}.
+   */
+  @Override
+  protected void storeMessageId(StructuredTableContext context, String messageId) throws Exception {
+    AppMetadataStore appMetadataStore = AppMetadataStore.create(context);
+    appMetadataStore.persistSubscriberState(getTopicId().getTopic(),
+                                            Constants.AuditLogging.AUDIT_LOG_WRITER_SUBSCRIBER, messageId);
+  }
+
+  /**
+   * Processes the give list of messages. This method will be called from the same transaction as
+   * the {@link #storeMessageId(StructuredTableContext, String)} call. If {@link Exception} is
+   * raised from this method, the messages as provided through the {@code messages} parameter will
+   * be replayed in the next call.
+   */
+  @Override
+  protected void processMessages(StructuredTableContext structuredTableContext,
+                                 Iterator<ImmutablePair<String, AuditLogContext>> messages) throws Exception {
+
+    Queue<AuditLogContext> auditLogContextQueue = new LinkedBlockingDeque<>();
```

Comment:
Any reason you need a blocking queue here? Seems like an intermediate buffer

Reply:
Do you mean using a simpler implementation like LinkedList. / ArrayDeque ?